hastory-cli/src/Hastory/Cli/OptParse.hs

```diff
@@ -1,62 +1,128 @@
+          DispatchGenGatherWrapperScript
+            GenGatherWrapperScriptSettings
+              {genGatherWrapperScriptSetRemoteInfo = mbRemoteStorageClientInfo}
+        CommandListRecentDirs ListRecentDirFlags {..} ->
           DispatchListRecentDirs
-            ListRecentDirSets {lrdSetBypassCache = fromMaybe False lrdArgBypassCache}
-        CommandChangeDir i -> DispatchChangeDir i
-        CommandGenChangeWrapperScript -> DispatchGenChangeWrapperScript
-        CommandSuggestAlias -> DispatchSuggestAlias
-    sets = do
+            ListRecentDirSettings {lrdSetBypassCache = fromMaybe False lrdArgBypassCache}
+        CommandChangeDir ChangeDirFlags {..} ->
+          DispatchChangeDir ChangeDirSettings {changeDirSetIdx = changeDirFlagsIdx}
+        CommandGenChangeWrapperScript _ ->
+          DispatchGenChangeWrapperScript GenChangeWrapperScriptSettings
+        CommandSuggestAlias _ -> DispatchSuggestAlias SuggestAliasSettings
+    getSettings = do
       home <- getHomeDir
       cacheDir <-
-        case flagCacheDir of
+        case flagCacheDir <|> envCacheDir <|> configCacheDir of
           Nothing -> resolveDir home ".hastory"
-          Just fcd -> resolveDir' fcd
-      let mbBaseUrl = parseBaseUrl . T.unpack =<< flagStorageServer
-      let mbRemoteStorageClientInfo =
-            RemoteStorageClientInfo <$> mbBaseUrl <*> flagStorageUsername <*> flagStoragePassword
+          Just cd -> resolveDir' cd
       pure Settings {setCacheDir = cacheDir, remoteStorageClientInfo = mbRemoteStorageClientInfo}
-
-getConfiguration :: Command -> Flags -> IO Configuration
-getConfiguration _ _ = pure Configuration
+    mbRemoteStorageClientInfo =
+      RemoteStorageClientInfo <$> (flagStorageServer <|> envStorageServer <|> configStorageServer) <*>
+      (flagStorageUsername <|> envStorageUsername <|> configStorageUsername) <*>
+      (flagStoragePassword <|> envStoragePassword <|> configStoragePassword)
+
+getConfiguration :: Path Abs File -> Flags -> Environment -> IO Configuration
+getConfiguration defaultConfigFile Flags {..} Environment {..} =
+  maybe useDefaultConfig useProvideConfigFile (flagConfigFile <|> envConfigFile)
+  where
+    useProvideConfigFile filePath = resolveFile' filePath >>= readConfigFile >>= displayingSchema
+    useDefaultConfig = do
+      fileExists <- doesFileExist defaultConfigFile
+      if fileExists
+        then readConfigFile defaultConfigFile >>= displayingSchema
+        else pure defaultConfiguration
+    displayingSchema :: Maybe Configuration -> IO Configuration
+    displayingSchema = maybe (die (T.unpack (prettySchemaDoc @Configuration))) pure
 
 getArguments :: IO Arguments
 getArguments = do
   args <- getArgs
   let result = runArgumentsParser args
   handleParseResult result
 
+getEnvironment :: IO Environment
+getEnvironment = Env.parse (Env.header "hastory") (Env.prefixed "HASTORY_" envParser)
+
+envParser :: Env.Parser Env.Error Environment
+envParser =
+  Env.prefixed
+    "HASTORY_"
```

Comment:
It looks like you use this prefix twice, please check by putting the help text in the output of `--help` (see the way the `intray-cli` does this).

Reply:
Indeed I did. Thank you.